### PR TITLE
Explicit import

### PR DESCRIPTION
**Pull request by matbesancon**
Following the discussion in #1051, making imports explicit whenever possible, and avoiding the blanket `using`

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -21,7 +21,7 @@ PDMats = "0.9"
 QuadGK = "2"
 SpecialFunctions = "0.8, 0.9"
 StatsBase = "0.32"
-StatsFuns = "0.8, 0.9"
+StatsFuns = ">= 0.9.3"
```
Comment (matbesancon) [inline Project.toml]: this is required by the use of a function in StatsFuns

**Comment by mschauer** [comment]
You also added some cosmetic improvements or is this just showing up because of the merge?

**Comment by matbesancon** [comment]
> You also added some cosmetic improvements or is this just showing up because of the merge?

What is showing up?

**Comment by matbesancon** [comment]
if related to the coverage diff, no idea why, all changes done are strictly equivalent

File: src/matrix/wishart.jl [hunk 2]
```diff
@@ -151,7 +151,7 @@ function _wishart_genA!(rng::AbstractRNG, p::Int, df::Real, A::AbstractMatrix)
     #               ~ Normal()                  when i > j
     #
     A .= zero(eltype(A))
-    for i = 1:p
+    for i in 1:p
```
Comment (mschauer) [inline src/matrix/wishart.jl]: @matbesancon Changes like these show up in my diff.

Comment (matbesancon) [reply src/matrix/wishart.jl]: Yes this was added you're right

**Comment by nalimilan** [comment]
I'm in favor of stopping unqualified `using X` statements (since breakage can happen when packages export new functions), but I'm quite reserved about qualifying all uses of functions. That's especially the case for SpecialFunctions, which is used all over the place. If you have a doubt it's very easy to find where it comes from by looking at the `using` or `import` line in src/Distributions.jl: that's another advantage of using explicit imports.

@ararslan Is that your reasoning as well?

File: Project.toml (hunk 1 above)
Comment (nalimilan) [reply Project.toml]: Then the new convention is this:
```suggestion
StatsFuns = "0.9.3"
```

CompatHelper is there to bump the dependency when a new version is tagged.

**Comment by matbesancon** [comment]
Thanks for your feedback, indeed for SpecialFunctions it might be better to do `using SF: f1, f2` instead, these are often the same ones coming back. I tried the qualified approach after the feedback on https://github.com/JuliaStats/Distributions.jl/pull/778, where people didn't seem enthusiastic about specified using

**Comment by nalimilan** [comment]
Let's hear what @ararslan and @simonbyrne say then.

**Comment by matbesancon** [comment]
I would be for `using SF: a, b` for things we have all over the place. For  others though, having one or two qualified imports in the code base helps a lot knowing where to look things up

**Comment by ararslan** [comment]
> @ararslan Is that your reasoning as well?

No, I'm a big proponent of simply `using X`. I especially dislike qualifying every call.

**Comment by nalimilan** [comment]
The problem with general `using X` is that if X starts exporting a new function, it can clash with other identifiers. For example, when Base started exporting `eachrow`, DataFrames had to adjust immediately to avoid breakage. That's really too bad as we have a very clear policy that minor releases shouldn't break dependencies... except if they use `using X`.

**Comment by ararslan** [comment]
Yeah there are cases where things like that happen, but I've found they occur sufficiently rarely in practice to not be much of a practical concern, and fixing/working around is quite straightforward.

**Comment by nalimilan** [comment]
Except if due to some complex dependency structure users can update the package which added the export, but not your package which fixes the bug. In that case they're screwed unless you make bugfix releases for all affected minor series. Granted that's theoretical, but with an expanding universe of packages, half-maintained ones will become more common and dependencies won't always be updated quickly.

**Comment by matbesancon** [comment]
The case was striking with the StatsFuns function that was available only in `>= 0.9.3`, with a blanket using, I had to google the function to see where it's supposed to be. With a specific or qualified using, I would have gotten an error message telling me StatsFuns with the current version does not have this or that

**Comment by matbesancon** [comment]
bump on this

File: src/multivariate/mvnormal.jl [hunk 3]
```diff
@@ -187,9 +187,9 @@ const IsoNormal  = MvNormal{Float64,ScalMat{Float64},Vector{Float64}}
 const DiagNormal = MvNormal{Float64,PDiagMat{Float64,Vector{Float64}},Vector{Float64}}
 const FullNormal = MvNormal{Float64,PDMat{Float64,Matrix{Float64}},Vector{Float64}}
 
-const ZeroMeanIsoNormal{Axes}  = MvNormal{Float64,ScalMat{Float64},Zeros{Float64,1,Axes}}
-const ZeroMeanDiagNormal{Axes} = MvNormal{Float64,PDiagMat{Float64,Vector{Float64}},Zeros{Float64,1,Axes}}
-const ZeroMeanFullNormal{Axes} = MvNormal{Float64,PDMat{Float64,Matrix{Float64}},Zeros{Float64,1,Axes}}
+const ZeroMeanIsoNormal{Axes}  = MvNormal{Float64,ScalMat{Float64},FillArrays.FillArrays.Zeros{Float64,1,Axes}}
```
Comment (nalimilan) [inline src/multivariate/mvnormal.jl]: Is the repetition intended?

Anyway, given how often `Zeros` is used, probably better import it.

File: src/multivariate/mvnormal.jl [hunk 4]
```diff
@@ -258,24 +258,25 @@ logdetcov(d::MvNormal) = logdet(d.Σ)
...
-    invquad!(r, d.Σ, x .- d.μ)
+    PDMats.invquad!(r, d.Σ, x .- d.μ)
 
 gradlogpdf(d::MvNormal, x::AbstractVector{<:Real}) = -(d.Σ \ (x .- d.μ))
 
 # Sampling (for GenericMvNormal)
 
-_rand!(rng::AbstractRNG, d::MvNormal, x::VecOrMat) =
-    add!(unwhiten!(d.Σ, randn!(rng, x)), d.μ)
+function _rand!(rng::AbstractRNG, d::MvNormal, x::VecOrMat)
```
Comment (nalimilan) [inline src/multivariate/mvnormal.jl]: The old one-line form seemed OK.

File: src/multivariate/mvlognormal.jl [hunk 5]
```diff
@@ -27,14 +27,17 @@ abstract type AbstractMvLogNormal <: ContinuousMultivariateDistribution end
 
 function insupport(::Type{D},x::AbstractVector{T}) where {T<:Real,D<:AbstractMvLogNormal}
     for i=1:length(x)
-      @inbounds 0.0<x[i]<Inf ? continue : (return false)
+        @inbounds 0.0<x[i]<Inf ? continue : (return false)
     end
-    true
+    return true
```
Comment (nalimilan) [inline src/multivariate/mvlognormal.jl]: This doesn't sound useful to me. Do you follow a particular style guide?

File: src/multivariate/mvnormal.jl (hunk 3 above)
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: agreed we use it a lot, done in last PR

File: src/multivariate/mvnormal.jl (hunk 4 above)
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: Ok, I considered awkward to have a one-liner spread on multiple lines, but not a drama, reverted in last commit

File: src/multivariate/mvlognormal.jl (hunk 5 above)
Comment (matbesancon) [reply src/multivariate/mvlognormal.jl]: both of these recently developed guides follow the convention of always using an explicit `return` keyword with long-form functions
https://github.com/jrevels/YASGuide
https://github.com/invenia/BlueStyle

File: src/multivariate/mvnormal.jl (hunk 4 above)
Comment (nalimilan) [reply src/multivariate/mvnormal.jl]: I interpret "one-liner" as referring to a single expression (i.e. no `;`). :-)

File: src/multivariate/mvnormal.jl [hunk 6]
```diff
@@ -333,22 +333,23 @@ end
 
 ## MLE estimation with covariance known
 
-fit_mle(g::MvNormalKnownCov{C}, ss::MvNormalKnownCovStats{C}) where {C<:AbstractPDMat} =
-    MvNormal(ss.sx * inv(ss.tw), g.Σ)
+function fit_mle(g::MvNormalKnownCov{C}, ss::MvNormalKnownCovStats{C}) where {C<:AbstractPDMat}
```
Comment (nalimilan) [inline src/multivariate/mvnormal.jl]: Keep one-liner too.

File: src/multivariate/mvlognormal.jl [hunk 7]
```diff
@@ -27,14 +27,17 @@ abstract type AbstractMvLogNormal <: ContinuousMultivariateDistribution end
...
 end
+
 insupport(l::AbstractMvLogNormal,x::AbstractVector{T}) where {T<:Real} = insupport(typeof(l),x)
 assertinsupport(::Type{D},m::AbstractVector) where {D<:AbstractMvLogNormal} = @assert insupport(D,m) "Mean of LogNormal distribution should be strictly positive"
 
-###Internal functions to calculate scale and location for a desired average and covariance
+"""
```
Comment (nalimilan) [inline src/multivariate/mvlognormal.jl]: This isn't appropriate as the comment described two functions, not one.

File: src/multivariate/mvlognormal.jl (hunk 5 above)
Comment (nalimilan) [reply src/multivariate/mvlognormal.jl]: OK. I had another convention in mind, which is that `return` must be used when the control flow is complex, i.e. the function doesn't just have a single return path. Not sure it's written anywhere.

Note that you haven't changed all occurrences (see e.g. `suffstats`). I'm not sure it makes sense to enforce this kind of conventino unless we enable an automated formatter like JuliaFormatter.

File: src/multivariate/mvlognormal.jl (hunk 7 above)
Comment (matbesancon) [reply src/multivariate/mvlognormal.jl]: reversed in last commit

File: src/multivariate/mvnormal.jl (hunk 6 above)
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: reversed in last commit

File: src/multivariate/mvnormal.jl (hunk 4 above)
Comment (matbesancon) [reply src/multivariate/mvnormal.jl]: yes got it :)

**Comment by matbesancon** [comment]
bump here

**Comment by mschauer** [comment]
I am not strongly in favour of either way to do it, but I am in favour of not introducing a hard rule. For example I like `using`, but I also see the risk of a new export messing up things with `using`. But then we can keep `using` packages we have somewhat control over, e.g. because they are looked over by the same people as `Distributions` and protect ourself by not `using` small packages from remote corners of the ecosystem. And those are also the packages where it is useful to qualify the function call, because at first sight it might not be clear where that particular function comes from.

**Comment by matbesancon** [comment]
Even for packages that are closer to Distributions, it can get tricky, see the bug caused by
`StatsFuns = "0.9.3"`, first reaction was to wonder where the `logmvbeta` function was coming from in the bug, with no better tool than googling it in hope of finding related packages

**Comment by matbesancon** [comment]
closing this for now, the PR is too old and most changes can be reproduced quickly. I still think blanket `using X` is a debugging burden for a package the size of Distributions

**Pull request closed without merging**